File: lib/core/src/main/java/io/loglens/core/Formatter.java
package io.loglens.core;

/**
 * 레코드 본문 조립기 — 계약의 emitter 쪽 구현.
 *
 * <pre>
 *   evt=&lt;EVENT&gt; k=v k=v | &lt;free message&gt;
 * </pre>
 *
 * <p>순수 문자열 로직이다. 안드로이드 의존이 전혀 없어서 일반 JVM 에서 그대로
 * 단위 테스트할 수 있고, 뷰어 파서와의 왕복 검증도 여기서 돌린다.
 */
public final class Formatter {

    /** 이 키로 넘긴 값은 필드가 아니라 {@code | } 뒤 자유 메시지가 된다. */
    public static final String MSG_KEY = "msg";

    private final Masker masker;
    private final int maxBytes;
    private final Diagnostics diag;

    /** 규격 위반을 개발자에게 알리는 통로. 릴리스에서는 무음. */
    public interface Diagnostics {
        void warn(String message);
        Diagnostics SILENT = m -> { };
    }

    public Formatter() {
        this(new Masker(), Truncator.DEFAULT_MAX_BYTES, Diagnostics.SILENT);
    }

    public Formatter(Masker masker, int maxBytes, Diagnostics diag) {
        this.masker = masker == null ? new Masker() : masker;
        this.maxBytes = maxBytes;
        this.diag = diag == null ? Diagnostics.SILENT : diag;
    }

    /**
     * 본문을 만든다.
     *
     * @param event 이벤트 코드 (대문자 SNAKE_CASE 권장)
     * @param kv    키-값 쌍의 varargs. 홀수면 마지막 키는 버리고 경고한다.
     * @param t     예외 (없으면 null). {@code err=} / {@code at=} 필드로 접힌다.
     */
    public String body(String event, Object[] kv, Throwable t) {
        StringBuilder sb = new StringBuilder(96);
        sb.append("evt=").append(Sanitizer.event(event));

        String freeMsg = null;

        if (kv != null && kv.length > 0) {
            if (kv.length % 2 != 0) {
                // 컴파일 타임에는 못 잡는다(varargs 의 대가). 개발 중에 시끄럽게 알린다.
                diag.warn("키-값 인자 개수가 홀수입니다 (" + kv.length + "). "
                        + "마지막 인자 '" + kv[kv.length - 1] + "' 를 버립니다. evt=" + event);
            }
            int pairs = (kv.length / 2) * 2;
            for (int i = 0; i < pairs; i += 2) {
                String key = kv[i] == null ? null : String.valueOf(kv[i]);
                if (!Sanitizer.isValidKey(key)) {
                    diag.warn("잘못된 키 '" + key + "' — 이 쌍을 건너뜁니다. evt=" + event);
                    continue;
                }
                if (MSG_KEY.equals(key)) {
                    freeMsg = kv[i + 1] == null ? null : String.valueOf(kv[i + 1]);
                    continue;
                }
                String value = masker.apply(key, Sanitizer.value(kv[i + 1]));
                sb.append(' ').append(key).append('=').append(value);
            }
        }

        if (t != null) {
            sb.append(' ').append("err=").append(Sanitizer.value(describe(t)));
            String frame = topFrame(t);
            if (frame != null) sb.append(' ').append("at=").append(Sanitizer.value(frame));
        }

        String msg = Sanitizer.message(freeMsg);
        if (msg != null) sb.append(" | ").append(msg);

        return Truncator.truncate(sb.toString(), maxBytes);
    }

    /** {@code ConnectException:timeout} 형태. 예외 메시지 원문을 통째로 덤프하지 않는다. */
    public static String describe(Throwable t) {
        String name = t.getClass().getSimpleName();
        String m = t.getMessage();
        if (m == null || m.isEmpty()) return name;
        // 메시지에 응답 본문이 통째로 실려오는 경우가 있다 (기획서 §9-3). 앞부분만.
        if (m.length() > 120) m = m.substring(0, 120);
        return name + ":" + m;
    }

    /** 앱 코드의 첫 스택 프레임. 한 줄 안에서 "어디서 터졌나"를 알려준다. */
    public static String topFrame(Throwable t) {
        StackTraceElement[] st = t.getStackTrace();
        if (st == null || st.length == 0) return null;
        StackTraceElement e = st[0];
        String cls = e.getClassName();
        int dot = cls.lastIndexOf('.');
        if (dot >= 0) cls = cls.substring(dot + 1);
        return cls + "." + e.getMethodName() + ":" + e.getLineNumber();
    }
}

File: lib/core/bin/main/io/loglens/core/Formatter.kt
package io.loglens.core

/**
 * 로그 본문을 조립합니다.
 *
 * ```
 * evt=<이벤트> 키=값 키=값 | <자유 메시지>
 * ```
 *
 * 순수 문자열 로직입니다. 안드로이드 의존이 전혀 없어서 일반 JVM 에서 그대로
 * 테스트할 수 있고, 뷰어 파서와 대조하는 검증도 여기서 돌립니다.
 */
class Formatter @JvmOverloads constructor(
    private val masker: Masker = Masker(),
    private val maxBytes: Int = Truncator.DEFAULT_MAX_BYTES,
    private val diag: Diagnostics = Diagnostics.SILENT,
) {

    /** 형식 위반을 개발자에게 알리는 통로입니다. 릴리스에서는 아무것도 하지 않습니다. */
    fun interface Diagnostics {
        fun warn(message: String)

        companion object {
            @JvmField
            val SILENT = Diagnostics { }
        }
    }

    /**
     * 본문을 만듭니다.
     *
     * @param event 이벤트 이름 (대문자 SNAKE_CASE 권장)
     * @param kv    키-값 쌍이 번갈아 들어옵니다. 개수가 홀수면 마지막 키는 버리고 경고합니다.
     * @param t     예외 (없으면 null). `err=` / `at=` 필드로 접혀 들어갑니다.
     */
    @JvmOverloads
    fun body(event: String?, kv: Array<out Any?>? = null, t: Throwable? = null): String {
        val sb = StringBuilder(96)
        sb.append("evt=").append(Sanitizer.event(event))

        var freeMsg: String? = null

        if (kv != null && kv.isNotEmpty()) {
            if (kv.size % 2 != 0) {
                // 컴파일 시점에는 못 잡습니다(가변 인자의 대가). 개발 중에 시끄럽게 알립니다.
                diag.warn(
                    "키-값 인자 개수가 홀수입니다 (${kv.size}). " +
                        "마지막 인자 '${kv.last()}' 를 버립니다. evt=$event"
                )
            }
            val pairs = (kv.size / 2) * 2
            var i = 0
            while (i < pairs) {
                val key = kv[i]?.toString()
                if (!Sanitizer.isValidKey(key)) {
                    diag.warn("잘못된 키 '$key' — 이 쌍을 건너뜁니다. evt=$event")
                    i += 2
                    continue
                }
                if (key == MSG_KEY) {
                    freeMsg = kv[i + 1]?.toString()
                    i += 2
                    continue
                }
                sb.append(' ').append(key).append('=')
                    .append(masker.apply(key, Sanitizer.value(kv[i + 1])))
                i += 2
            }
        }

        if (t != null) {
            sb.append(' ').append("err=").append(Sanitizer.value(describe(t)))
            topFrame(t)?.let { sb.append(' ').append("at=").append(Sanitizer.value(it)) }
        }

        Sanitizer.message(freeMsg)?.let { sb.append(" | ").append(it) }

        return Truncator.truncate(sb.toString(), maxBytes)!!
    }

    companion object {
        /** 이 키로 넘긴 값은 필드가 아니라 `| ` 뒤의 자유 메시지가 됩니다. */
        const val MSG_KEY = "msg"

        /** `ConnectException:timeout` 형태. 예외 메시지를 통째로 덤프하지 않습니다. */
        @JvmStatic
        fun describe(t: Throwable): String {
            val name = t.javaClass.simpleName
            var m = t.message
            if (m.isNullOrEmpty()) return name
            // 예외 메시지에 응답 본문이 통째로 실려 오는 경우가 있습니다. 앞부분만 씁니다.
            if (m.length > 120) m = m.substring(0, 120)
            return "$name:$m"
        }

        /** 스택의 첫 프레임. 한 줄 안에서 "어디서 터졌나"를 알려 줍니다. */
        @JvmStatic
        fun topFrame(t: Throwable): String? {
            val e = t.stackTrace?.firstOrNull() ?: return null
            val cls = e.className.substringAfterLast('.')
            return "$cls.${e.methodName}:${e.lineNumber}"
        }
    }
}
